index.d.ts

```diff
@@ -0,0 +1,110 @@
+import { Action, ActionCreator, AnyAction, StoreEnhancer, Store } from 'redux';
+
+export interface StoreCreator {
+  <S, A extends Action>(
+    reducer: LoopReducer<S, A>,
+    preloadedState: S,
+    enhancer: StoreEnhancer<S>
+  ): Store<S>;
+}
+
+export type Loop<S, A extends Action> = [S, CmdType<A>];
+
+export interface LoopReducer<S, A extends Action> {
+  (state: S | undefined, action: A): S | Loop<S, A>;
```

Comment:
@willisplummer where i'm still confused about the generics on the Action types is here. A is used not just in the return type, but in the parameter for the Action. But all reducers need to be able to take any action type, not just a subset. Should this line be `(state: S | undefined, action: AnyAction): S | Loop<S, A>;`

Reply:
@bdwain that makes sense to only include the generic in the return type I think.

I think this stackexchange question was the origin of this typing and it seems like we still get the safety I was after with `action: AnyAction`

https://stackoverflow.com/questions/45899039/how-do-i-write-a-function-type-that-strictly-requires-its-parameters-in-typescri/45900816#45900816

Nice suggestion 👍